### PR TITLE
Improved ReadTheDocs Documentation

### PR DESCRIPTION
**Pull request by Myoldmopar**
Pull request overview
---------------------
A notable cleanup on our Sphinx-based documentation.
 - Improved EpJSON schema visualization, with badges for required fields, required objects, and field types
 - Improved theming, much more modern RTD-style appearance
 - Added JS filter for quickly finding the object of interest
 - Other conf cleanup.

Documentation with the changes in this branch can be found here: https://energyplus.readthedocs.io/en/improveschemadocs/

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: .readthedocs.yml [hunk 1]
```diff
@@ -2,3 +2,9 @@ version: 2
 
 sphinx:
   configuration: doc/readthedocs/sphinx/conf.py
+
+# read the docs image build already includes sphinx-rtd-theme, so this isn't needed
+# however, keeping it here and running it on RTD is a near zero cost operation and makes sure the requirements are ok
+python:
+  install:
+    - requirements: doc/readthedocs/requirements.txt
```
Comment (Myoldmopar) [inline .readthedocs.yml:10]: Comment explains this, but basically, if you are building the sphinx docs as a dev, install the reqs.  RTD doesn't actually need this, but it doesn't hurt anything.

File: doc/readthedocs/sphinx/.templates/footer.html [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{% extends "!footer.html" %}
+{%- block contentinfo %}
+        <a href="https://energyplus.net">EnergyPlus Web Site</a> <br>
+        <a href="https://github.com/NREL/EnergyPlus">EnergyPlus Github Repo</a> <br>
+        <a href="https://energyplus.readthedocs.io/en/latest/">API Documentation Home</a> <br>
+{%- endblock %}
```
Comment (Myoldmopar) [inline doc/readthedocs/sphinx/.templates/footer.html:6]: Override the footer template to add a few some links.

File: doc/readthedocs/sphinx/_static/main.css [hunk 3]
```diff
@@ -0,0 +1,7 @@
+.form__field {
+    width: 100%;
+    font-size: 1.3rem;
+    padding: 7px 7px;
+    margin-bottom: 10px;
+    background: transparent;
+}
```
Comment (Myoldmopar) [inline doc/readthedocs/sphinx/_static/main.css:7]: Minor tweaking of object filter box.

File: doc/readthedocs/sphinx/_static/main.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            .keyup(function () {
+                // fire the above change event after every letter
+                $(this).change();
+            });
+    }
+
+    $(function () {
+        listFilter();
+    });
+}(jQuery));
```
Comment (Myoldmopar) [inline doc/readthedocs/sphinx/_static/main.js:33]: Little bit of JS to allow filtering the object list.  I'm not a JS master, but it works well and is moderately clear.

File: doc/readthedocs/sphinx/conf.py [hunk 5]
```diff
@@ -1,32 +1,73 @@
...
+}
+# html_sidebars = {
+#     '**': [
+#         'about.html',
+#         'navigation.html',
+#         'relations.html',
+#         'searchbox.html',
+#     ]
+# }
+
```
Comment (Myoldmopar) [inline doc/readthedocs/sphinx/conf.py:69]: A lot of these changes are just rearranging, but I am also specifying some extra files to bring in during build.

File: doc/readthedocs/sphinx/conf.py [hunk 6]
```diff
@@ -142,25 +193,56 @@
...
+                elif auto_calculate_found and numeric_found:
+                    field_type = 'auto_calculate_numeric'
+                elif auto_size_found and integer_found:
+                    field_type = 'auto_size_integer'
+                elif numeric_found:  # this handles goofy things like RefrigerationSystem:NumCompStages
+                    field_type = 'number'
+            this_field_type = field_type_icon[field_type]
+        req_prefix = f":abbr:`🅁 (Required Field)` " if field_name in required_fields else ""
+        type_prefix = this_field_type
+        details += f"* {req_prefix}{type_prefix} `{field_name}` {default_string}\n"
```
Comment (Myoldmopar) [inline doc/readthedocs/sphinx/conf.py:235]: In this section I'm more carefully determining field data to present it better.

File: doc/readthedocs/sphinx/in.schema.rst.in [hunk 7]
```diff
@@ -1,8 +1,22 @@
...
 
+Icon Legend:
+
+* 🅁 This field, or object, is a required input
+* Ⓝ Numeric Field (Floating point usually)
+* Ⓘ Integer Value
+* Ⓢ String Input
+* Ⓧ Array Field, which will contain an extensible amount of groups of fields
+* ⒶⓃ Auto-sizable, or Auto-calculable, floating point input field
+* ⒶⒾ Auto-sizable, or Auto-calculable, integer value
```
Comment (Myoldmopar) [inline doc/readthedocs/sphinx/in.schema.rst.in:16]: Added a legend to the top of the schema doc.  Looks pretty good.

File: release/favicon_white_plus.png
Comment (Myoldmopar) [inline release/favicon_white_plus.png:1]: Added an alternative icon with the plus logo in white.  The blue was completely invisible on the template so I had to do something.

**Review by Myoldmopar — COMMENTED** [review]
Simple changes, anyone care to review/comment?

**Comment by Myoldmopar** [comment]
Ended up trying to do a little more and it came out awesome with almost no effort...  https://energyplus.readthedocs.io/en/improveschemadocs/tips_and_tricks/tips_and_tricks.html

**Comment by rraustad** [comment]
Looking at the result, this does look very good. Is it possible to have a "table of contents" line next to each section title that takes you back to the table of contents?

I don't know what this permalink thing is but could a hover somehow get back to TOC?

image

**Comment by Myoldmopar** [comment]
So the permalink is just a way to provide a user a URL to that exact point.  Simply an anchor.  As for getting back to the TOC, I think you could click the page title on the left bar and it would go back to the top, or press "home" on the keyboard to get back to the top.  All that said, I think I can do a little work to add a link to top that stays on the page.  I'll check it out.

**Comment by Myoldmopar** [comment]
How about a little link at the bottom that stays there so you can always get back to the top?  https://energyplus.readthedocs.io/en/improveschemadocs/tips_and_tricks/tips_and_tricks.html

**Comment by Myoldmopar** [comment]
@jasondegraw https://energyplus.readthedocs.io/en/improveschemadocs/quick_start/quick_start.html

OK, so read the docs seems to have made some change that caused many, many users to quickly insert a workaround to get a specific version of urllib3.  I hadn't noticed that docs for our develop branch failed to build unexpectedly with the latest commit, primarily because we hadn't made any related change to cause it.  Anyway, I did the same workaround to fix it, and that fix is included in this branch.  As such, I want to merge this in quickly to get our develop branch building again.  Countdown has begun.

**Pull request merged**